[Time] Fix tm_mon starting at 0 and tm_year at 1900

TD-er opened the pull request:
All over the code there were incorrect assumptions about the offsets in `struct tm`.

See [struct tm](https://www.cplusplus.com/reference/ctime/tm/)
![image](https://user-images.githubusercontent.com/3751318/106747373-1e693800-6624-11eb-88db-655a69d26ea2.png)

Closes: #3441 
Fixes: #3440

TD-er committed:
[Time] Fix tm_mon starting at 0 and tm_year at 1900

All over the code there were incorrect assumptions about the offsets in `struct tm`

TD-er commented:
Test build:
- [ESPEasy_ESP32_mega-20210114-51-PR_3487.zip](https://www.dropbox.com/s/59axkctluesvqhz/ESPEasy_ESP32_mega-20210114-51-PR_3487.zip?dl=0)
- [ESPEasy_ESP82xx_mega-20210114-51-PR_3487.zip](https://www.dropbox.com/s/7dduzg9woht155p/ESPEasy_ESP82xx_mega-20210114-51-PR_3487.zip?dl=0)

@clumsy-stefan and @uzi18 can you test it too?

clumsy-stefan commented:
I'm happy to test, but can't use the test builds, as i have my own set of plugins. so I guess I need to merge PR 3487 into my current branch. Iis that all or are there other fixes in other PR's /commits?

TD-er commented:
This PR is based on the current HEAD of mega.
If you're already working on the HEAD of mega, then this is just what it is, only fixes in time calculation.

TD-er commented:
By the way, if this is now fixed and no really urgent issues emerge in the next few days, I think this can be considered a RC :)

TD-er committed:
[Time] Cleanup time calculation code

clumsy-stefan commented:
I did my build now and installed it on a few test-nodes.... if they run fine for the day, I'll update the rest of my nodes and see what happens... will report then....

clumsy-stefan commented:
up to now everything runs fine, one thing I noticed (though not sure if it was already happening before), in the cron plugin when I use a timecode like: `14 * * * * *` the execution runs every 68 seconds (and not 14 seconds after full minute)...

TD-er commented:
> up to now everything runs fine, one thing I noticed (though not sure if it was already happening before), in the cron plugin when I use a timecode like: `14 * * * * *` the execution runs every 68 seconds (and not 14 seconds after full minute)...

Can you test it with sending `%systime% to the logs as logentry?

clumsy-stefan commented:
Hmm... the log looks even more strange, it seems to run every second until the time is reached (eg. sec. 14) and then again after about 30 sec.... until time reached again....

```
11176887: CCS811 : tVOC: 26, eCO2: 574
11186890: CCS811 : tVOC: 26, eCO2: 574
11186974: EVENT: Cron#Crontest
11187043: ACT  : event,date:15:04:32
11187044: EVENT: date:15:04:32
11187974: EVENT: Cron#Crontest
11188048: ACT  : event,date:15:04:33
11188050: EVENT: date:15:04:33
11188974: EVENT: Cron#Crontest
11189045: ACT  : event,date:15:04:34
11189046: EVENT: date:15:04:34
11189969: EVENT: Cron#Crontest
11190042: ACT  : event,date:15:04:35
11190044: EVENT: date:15:04:35
11190970: EVENT: Cron#Crontest
11191043: ACT  : event,date:15:04:36
11191045: EVENT: date:15:04:36
11191969: EVENT: Cron#Crontest
11192042: ACT  : event,date:15:04:37
11192043: EVENT: date:15:04:37
11192974: EVENT: Cron#Crontest
11193044: ACT  : event,date:15:04:38
11193046: EVENT: date:15:04:38
11193435: WD   : Uptime 187 ConnectFailures 0 FreeMem 6808 WiFiStatus WL_CONNECTED ESPeasy internal wifi status: Conn. IP Init
11193969: EVENT: Cron#Crontest
11194043: ACT  : event,date:15:04:39
11194044: EVENT: date:15:04:39
11194969: EVENT: Cron#Crontest
11195039: ACT  : event,date:15:04:40
11195041: EVENT: date:15:04:40
11195974: EVENT: Cron#Crontest
11196047: ACT  : event,date:15:04:41
11196048: EVENT: date:15:04:41
11196941: CCS811 : tVOC: 25, eCO2: 566
11196974: EVENT: Cron#Crontest
11197045: ACT  : event,date:15:04:42
11197047: EVENT: date:15:04:42
11197973: EVENT: Cron#Crontest
11198047: ACT  : event,date:15:04:43
11198049: EVENT: date:15:04:43
11198969: EVENT: Cron#Crontest
11199041: ACT  : event,date:15:04:44
11199042: EVENT: date:15:04:44
11199975: EVENT: Cron#Crontest
11200047: ACT  : event,date:15:04:45
11200048: EVENT: date:15:04:45
11200969: EVENT: Cron#Crontest
11201041: ACT  : event,date:15:04:46
11201043: EVENT: date:15:04:46
11201969: EVENT: Cron#Crontest
11202042: ACT  : event,date:15:04:47
11202043: EVENT: date:15:04:47
11202973: EVENT: Cron#Crontest
11203043: ACT  : event,date:15:04:48
11203044: EVENT: date:15:04:48
11203972: EVENT: Cron#Crontest
11204044: ACT  : event,date:15:04:49
11204045: EVENT: date:15:04:49
11204972: EVENT: Cron#Crontest
11205045: ACT  : event,date:15:04:50
11205047: EVENT: date:15:04:50
11205979: EVENT: Cron#Crontest
11206052: ACT  : event,date:15:04:51
11206054: EVENT: date:15:04:51
11206887: CCS811 : tVOC: 26, eCO2: 574
11206980: EVENT: Cron#Crontest
11207053: ACT  : event,date:15:04:52
11207055: EVENT: date:15:04:52
11207979: EVENT: Cron#Crontest
11208049: ACT  : event,date:15:04:53
11208051: EVENT: date:15:04:53
11208974: EVENT: Cron#Crontest
11209044: ACT  : event,date:15:04:54
11209045: EVENT: date:15:04:54
11209974: EVENT: Cron#Crontest
11210044: ACT  : event,date:15:04:55
11210045: EVENT: date:15:04:55
11210968: EVENT: Cron#Crontest
11211041: ACT  : event,date:15:04:56
11211043: EVENT: date:15:04:56
11211968: EVENT: Cron#Crontest
11212040: ACT  : event,date:15:04:57
11212042: EVENT: date:15:04:57
11212974: EVENT: Cron#Crontest
11213048: ACT  : event,date:15:04:58
11213049: EVENT: date:15:04:58
11213976: EVENT: Cron#Crontest
11214048: ACT  : event,date:15:04:59
11214050: EVENT: date:15:04:59
11214939: EVENT: Clock#Time=Thu,15:05
11215049: EVENT: Cron#Crontest
11215121: ACT  : event,date:15:05:01
11215124: EVENT: date:15:05:01
11215968: EVENT: Cron#Crontest
11216039: ACT  : event,date:15:05:01
11216040: EVENT: date:15:05:01
11216888: CCS811 : tVOC: 26, eCO2: 574
11216890: BME280: Apply temp offset -2.00C humidity 32.68% => 34.68% temperature 28.19C => 26.19C dew point 9.38C
11216907: BME280 : Address: 0x76
11216908: BME280 : Temperature: 26
11216909: BME280 : Humidity: 35
11216910: BME280 : Barometric Pressure: 1014
11216977: EVENT: Cron#Crontest
11217050: ACT  : event,date:15:05:02
11217052: EVENT: date:15:05:02
11217201: EVENT: bme280#temperature=26
11217303: EVENT: bme280#humidity=35
11217406: EVENT: bme280#pressure=1014
11217969: EVENT: Cron#Crontest
11218042: ACT  : event,date:15:05:03
11218044: EVENT: date:15:05:03
11218968: EVENT: Cron#Crontest
11219038: ACT  : event,date:15:05:04
11219039: EVENT: date:15:05:04
11219974: EVENT: Cron#Crontest
11220047: ACT  : event,date:15:05:05
11220048: EVENT: date:15:05:05
11220620: SW   : State 0.00
11220680: EVENT: ch1#switch=0
11220974: EVENT: Cron#Crontest
11221046: ACT  : event,date:15:05:06
11221048: EVENT: date:15:05:06
11221969: EVENT: Cron#Crontest
11222043: ACT  : event,date:15:05:07
11222045: EVENT: date:15:05:07
11222968: EVENT: Cron#Crontest
11223041: ACT  : event,date:15:05:08
11223043: EVENT: date:15:05:08
11223433: WD   : Uptime 187 ConnectFailures 0 FreeMem 6616 WiFiStatus WL_CONNECTED ESPeasy internal wifi status: Conn. IP Init
11223968: EVENT: Cron#Crontest
11224042: ACT  : event,date:15:05:09
11224043: EVENT: date:15:05:09
11224200: EVENT: Crontest#LastExecution=1612451072.00
11224301: EVENT: Crontest#NextExecution=1612451072.00
11224968: EVENT: Cron#Crontest
11225039: ACT  : event,date:15:05:10
11225040: EVENT: date:15:05:10
11225974: EVENT: Cron#Crontest
11226048: ACT  : event,date:15:05:11
11226049: EVENT: date:15:05:11
11226938: CCS811 : tVOC: 26, eCO2: 574
11226969: EVENT: Cron#Crontest
11227041: ACT  : event,date:15:05:12
11227043: EVENT: date:15:05:12
11227968: EVENT: Cron#Crontest
11228043: ACT  : event,date:15:05:13
11228044: EVENT: date:15:05:13
11228976: EVENT: Cron#Crontest
11229048: ACT  : event,date:15:05:14
11229050: EVENT: date:15:05:14
11236890: CCS811 : tVOC: 25, eCO2: 566
11246888: CCS811 : tVOC: 26, eCO2: 574
11253435: WD   : Uptime 188 ConnectFailures 0 FreeMem 6704 WiFiStatus WL_CONNECTED ESPeasy internal wifi status: Conn. IP Init
11256929: CCS811 : tVOC: 26, eCO2: 574
11266889: CCS811 : tVOC: 26, eCO2: 574
11274935: EVENT: Clock#Time=Thu,15:06
11276891: CCS811 : tVOC: 25, eCO2: 566
11280621: SW   : State 0.00
11280676: EVENT: ch1#switch=0
11283436: WD   : Uptime 188 ConnectFailures 0 FreeMem 6488 WiFiStatus WL_CONNECTED ESPeasy internal wifi status: Conn. IP Init
11284082: EVENT: Crontest#LastExecution=1612451072.00
11284180: EVENT: Crontest#NextExecution=1612451200.00
11286935: CCS811 : tVOC: 25, eCO2: 566
11296887: CCS811 : tVOC: 26, eCO2: 574
11306884: CCS811 : tVOC: 26, eCO2: 574
11313436: WD   : Uptime 189 ConnectFailures 0 FreeMem 6376 WiFiStatus WL_CONNECTED ESPeasy internal wifi status: Conn. IP Init
11314973: EVENT: Cron#Crontest
11315043: ACT  : event,date:15:06:40
11315044: EVENT: date:15:06:40
11315969: EVENT: Cron#Crontest
11316042: ACT  : event,date:15:06:41
11316043: EVENT: date:15:06:41
11316948: CCS811 : tVOC: 26, eCO2: 574
11316972: EVENT: Cron#Crontest
11317044: ACT  : event,date:15:06:42
11317045: EVENT: date:15:06:42
11317969: EVENT: Cron#Crontest
11318042: ACT  : event,date:15:06:43
11318044: EVENT: date:15:06:43
11318970: EVENT: Cron#Crontest
11319042: ACT  : event,date:15:06:44
11319043: EVENT: date:15:06:44
11319969: EVENT: Cron#Crontest
11320044: ACT  : event,date:15:06:45
11320045: EVENT: date:15:06:45
11320969: EVENT: Cron#Crontest
11321042: ACT  : event,date:15:06:46
11321043: EVENT: date:15:06:46
11321969: EVENT: Cron#Crontest
11322042: ACT  : event,date:15:06:47
11322044: EVENT: date:15:06:47
11322969: EVENT: Cron#Crontest
11323038: ACT  : event,date:15:06:48
11323039: EVENT: date:15:06:48
11323974: EVENT: Cron#Crontest
11324046: ACT  : event,date:15:06:49
11324048: EVENT: date:15:06:49
11324976: EVENT: Cron#Crontest
11325049: ACT  : event,date:15:06:50
11325051: EVENT: date:15:06:50
11325979: EVENT: Cron#Crontest
11326053: ACT  : event,date:15:06:51
11326054: EVENT: date:15:06:51
11326888: CCS811 : tVOC: 26, eCO2: 574
11326974: EVENT: Cron#Crontest
11327046: ACT  : event,date:15:06:52
11327048: EVENT: date:15:06:52
11327975: EVENT: Cron#Crontest
11328044: ACT  : event,date:15:06:53
11328045: EVENT: date:15:06:53
11328969: EVENT: Cron#Crontest
11329039: ACT  : event,date:15:06:54
11329040: EVENT: date:15:06:54
11329968: EVENT: Cron#Crontest
11330038: ACT  : event,date:15:06:55
11330039: EVENT: date:15:06:55
11330997: EVENT: Cron#Crontest
11331070: ACT  : event,date:15:06:56
11331072: EVENT: date:15:06:56
11331967: EVENT: Cron#Crontest
11332039: ACT  : event,date:15:06:57
11332041: EVENT: date:15:06:57
11332968: EVENT: Cron#Crontest
11333042: ACT  : event,date:15:06:58
11333044: EVENT: date:15:06:58
11333969: EVENT: Cron#Crontest
11334041: ACT  : event,date:15:06:59
11334043: EVENT: date:15:06:59
11334944: EVENT: Clock#Time=Thu,15:07
11335053: EVENT: Cron#Crontest
11335124: ACT  : event,date:15:07:01
11335126: EVENT: date:15:07:01
11335969: EVENT: Cron#Crontest
11336038: ACT  : event,date:15:07:01
11336040: EVENT: date:15:07:01
11336891: CCS811 : tVOC: 26, eCO2: 574
11336974: EVENT: Cron#Crontest
11337044: ACT  : event,date:15:07:02
11337046: EVENT: date:15:07:02
11337969: EVENT: Cron#Crontest
11338041: ACT  : event,date:15:07:03
11338043: EVENT: date:15:07:03
11338968: EVENT: Cron#Crontest
11339038: ACT  : event,date:15:07:04
11339040: EVENT: date:15:07:04
11339974: EVENT: Cron#Crontest
11340046: ACT  : event,date:15:07:05
11340048: EVENT: date:15:07:05
11340613: SW   : State 0.00
11340673: EVENT: ch1#switch=0
11340974: EVENT: Cron#Crontest
11341047: ACT  : event,date:15:07:06
11341048: EVENT: date:15:07:06
11341973: EVENT: Cron#Crontest
11342046: ACT  : event,date:15:07:07
11342047: EVENT: date:15:07:07
11342969: EVENT: Cron#Crontest
11343041: ACT  : event,date:15:07:08
11343043: EVENT: date:15:07:08
11343433: WD   : Uptime 189 ConnectFailures 0 FreeMem 6088 WiFiStatus WL_CONNECTED ESPeasy internal wifi status: Conn. IP Init
11343969: EVENT: Cron#Crontest
11344041: ACT  : event,date:15:07:09
11344043: EVENT: date:15:07:09
11344173: EVENT: Crontest#LastExecution=1612451200.00
11344273: EVENT: Crontest#NextExecution=1612451200.00
11344977: EVENT: Cron#Crontest
11345047: ACT  : event,date:15:07:10
11345048: EVENT: date:15:07:10
11345974: EVENT: Cron#Crontest
11346048: ACT  : event,date:15:07:11
11346049: EVENT: date:15:07:11
11346940: CCS811 : tVOC: 26, eCO2: 574
11346974: EVENT: Cron#Crontest
11347045: ACT  : event,date:15:07:12
11347047: EVENT: date:15:07:12
11347969: EVENT: Cron#Crontest
11348043: ACT  : event,date:15:07:13
11348044: EVENT: date:15:07:13
11348969: EVENT: Cron#Crontest
11349042: ACT  : event,date:15:07:14
11349044: EVENT: date:15:07:14
11356889: CCS811 : tVOC: 26, eCO2: 574
```

TD-er commented:
What do you have in the rules for this?
I can't explain the extra events just yet.
But this seems like it is still a resolution issue as you tried to fix in your PR:

```
11344173: EVENT: Crontest#LastExecution=1612451200.00
11344273: EVENT: Crontest#NextExecution=1612451200.00
```

clumsy-stefan commented:
Rule:
```
on Cron#Crontest do
  event,date:%systime%
endon
```

clumsy-stefan commented:
@TD-er as per your last comment about the resolution issue, I compiled a version including my 32/64 bit fix/PR. with this the log looks way better, so probably this  could be the issue at all?

Also execution always happens exactly once at the "correct" time.

EDIT: one more thing I just realised: the high byte is always zero... so probably it's enough to store it in 32 bit just need to make sure it's converted to 64bit before pushing it in the time struct..

Log after including my PR:
```
86701: CCS811 : tVOC: 0, eCO2: 405
96697: CCS811 : tVOC: 0, eCO2: 405
106250: WD   : Uptime 2 ConnectFailures 0 FreeMem 7136 WiFiStatus WL_CONNECTED ESPeasy internal wifi status: Conn. IP Init
106700: CCS811 : tVOC: 0, eCO2: 405
115696: EVENT: Clock#Time=Thu,20:08
116694: CCS811 : tVOC: 0, eCO2: 405
126700: CCS811 : tVOC: 0, eCO2: 405
129738: EVENT: Cron#Crontest
129832: ACT  : event,date:20:08:14
129834: EVENT: date:20:08:14
134179: SW   : State 0.00
134237: EVENT: ch1#switch=0
136244: WD   : Uptime 2 ConnectFailures 0 FreeMem 6448 WiFiStatus WL_CONNECTED ESPeasy internal wifi status: Conn. IP Init
136638: EVENT: Crontest#LastExecution_L=45058716581981332893.74
136749: CCS811 : tVOC: 0, eCO2: 405
136757: EVENT: Crontest#LastExecution_H=0.00
136854: EVENT: Crontest#NextExecution_L=45058980464772000473.77
136959: EVENT: Crontest#NextExecution_H=0.00
146702: CCS811 : tVOC: 0, eCO2: 405
156700: CCS811 : tVOC: 0, eCO2: 405
166250: WD   : Uptime 3 ConnectFailures 0 FreeMem 6344 WiFiStatus WL_CONNECTED ESPeasy internal wifi status: Conn. IP Init
166700: CCS811 : tVOC: 0, eCO2: 405
175696: EVENT: Clock#Time=Thu,20:09
176700: CCS811 : tVOC: 1, eCO2: 408
186699: CCS811 : tVOC: 0, eCO2: 405
189737: EVENT: Cron#Crontest
189812: ACT  : event,date:20:09:14
189814: EVENT: date:20:09:14
194180: SW   : State 0.00
194236: EVENT: ch1#switch=0
196256: WD   : Uptime 3 ConnectFailures 0 FreeMem 5920 WiFiStatus WL_CONNECTED ESPeasy internal wifi status: Conn. IP Init
196642: EVENT: Crontest#LastExecution_L=45058980464772000473.77
196777: CCS811 : tVOC: 0, eCO2: 405
196785: EVENT: Crontest#LastExecution_H=0.00
196883: EVENT: Crontest#NextExecution_L=45059244347562668053.79
196990: EVENT: Crontest#NextExecution_H=0.00
206701: CCS811 : tVOC: 0, eCO2: 405
216701: CCS811 : tVOC: 0, eCO2: 405
226250: WD   : Uptime 4 ConnectFailures 0 FreeMem 6072 WiFiStatus WL_CONNECTED ESPeasy internal wifi status: Conn. IP Init
226700: CCS811 : tVOC: 0, eCO2: 405
235696: EVENT: Clock#Time=Thu,20:10
236700: CCS811 : tVOC: 1, eCO2: 408
246700: CCS811 : tVOC: 0, eCO2: 405
249711: EVENT: Cron#Crontest
249785: ACT  : event,date:20:10:14
249787: EVENT: date:20:10:14
254184: SW   : State 0.00
254236: EVENT: ch1#switch=0
256251: WD   : Uptime 4 ConnectFailures 0 FreeMem 5696 WiFiStatus WL_CONNECTED ESPeasy internal wifi status: Conn. IP Init
256700: CCS811 : tVOC: 0, eCO2: 405
256707: EVENT: Crontest#LastExecution_L=45059244347562668053.79
256808: EVENT: Crontest#LastExecution_H=0.00
257010: EVENT: Crontest#NextExecution_L=45059508230353326752.03
257111: EVENT: Crontest#NextExecution_H=0.00
```

clumsy-stefan commented:
Also just tried manual time settings for december (where I had the original issue, showing a month of "13" and with january, both seem to work correctly...

![image](https://user-images.githubusercontent.com/32835100/106945547-3aee9880-6728-11eb-8660-517649d6d560.png)

![image](https://user-images.githubusercontent.com/32835100/106945698-6a9da080-6728-11eb-87f8-c0e9dce3bf5c.png)

TD-er commented:
Check, good to know it is indeed a resolution issue for Cron.

@tonhuisman Did you already start on a universal interface to `UserVar`.
If not, then I would like to make a start for it for this issue with Cron.

tonhuisman commented:
> @tonhuisman Did you already start on a universal interface to `UserVar`.
> If not, then I would like to make a start for it for this issue with Cron.

 I haven't started anything in that direction, so please go ahead 👍

clumsy-stefan commented:
my test units ran fine over night, so I updated all my units with my latest version now... hope they'll survive ;) I'll report back later today/tonight...

clumsy-stefan commented:
On units with pcfgpio's attached I have a issues with this version, not sure why this is related to the changes. but setting an output port to 0 just keeps it at 1:

![image](https://user-images.githubusercontent.com/32835100/107000681-92245580-6788-11eb-9e55-37000fbd691c.png)

The log shows the same behaviour, I can't set the port to 0 anymore...

TD-er commented:
Judging your "ort#" in the output, you're not running the latest code for Commands/GPIO.cpp

clumsy-stefan commented:
Hmm.. just remerged the latest mega into my test-branch and still get the same issue. But I had a working version from Feb. 2nd, so the problem had to be introduced after that.

So I did a new version with commit #8a6dc87b4bda4fe32fec6dedf8911aa063be8f66 reverted, and it worked again... so it seems to be that `[Rules] Make parsing numerical a bit more robust` has some weird flaw in it...

Still, also without this commit the "p" is missing (but setting works again):
![image](https://user-images.githubusercontent.com/32835100/107035060-e135af00-67b7-11eb-9a9e-5f1e476255ab.png)

TD-er commented:
> So I did a new version with commit #8a6dc87b4bda4fe32fec6dedf8911aa063be8f66 reverted, and it worked again... so it seems to be that `[Rules] Make parsing numerical a bit more robust` has some weird flaw in it...

OK, will check the changes in there to see if I messed up somewhere

TD-er commented:
Just checked on a node here, sending `pcfgpio,5,0`
```
{
"log": "ort#5: set to -1",
"plugin": 19,
"pin": 5,
"mode": "offline",
"state": -1
}
```
So maybe it is just a matter of an extra -1?
At least I do see the same "ort#5" as you saw.

Edit:
Found the reason for the missing part.
```c++

    String reply;
    reply.reserve(128);
    reply += F("{\n\"log\": \"");
    reply += log.substring(7, 32); // truncate to 25 chars, max MQTT message size = 128 including header...
    reply += F("\",\n\"plugin\": ");
```
It does use a fixed offset

clumsy-stefan commented:
I think the "-1" as result comes because you have not actually connected a pcf module... -1 means "not conencted" or "no communication" or so...

clumsy-stefan commented:
I created a new PR for the 64bit time_t issue which is a lot simpler as it seems the high byte of the combined value is always 0x00. I tested this version for a number of cases and it seems to work (no complete warranty though). But like this the plugin still only uses 2 user variables (instead of 4). PR #3490

clumsy-stefan commented:
quick update: all my nodes (~30 with variety of sensors and modules) ran now fine for something over 12 hours. The build includes PR #3487 (this time fix) and #3490 (cron 64 bit fix) with reverted commit #8a6dc87b4bda4fe32fec6dedf8911aa063be8f66 .

TD-er committed:
[Cron] Fix cron resolution + value updates

TD-er commented:
@clumsy-stefan Can you test the code of this PR?
Right now I do have a lot of internet connection issues as I'm having a major internet outage (already started yesterday) and thus we're running with laptops on my mobile as hotspot.

At least it makes it hard for me to make a test build so I hope you can manage.
I can build one on my laptop, but may take a while.

What I did change is that I changed the union used in the cron plugin to make sure it is using a 32 bit data type as `time_t` may be 8 bytes on one system and 4 bytes on another, which may cause lots of issues.
So I really would like you can build it on your system as the initial code did not result in strange issues as described by you.
If it still is causing erratic behavior, I will add attributes to make sure data will be properly aligned (or just make a conversion object that will work regardless the compile platform)

I also added a new plugin function to fetch a formatted variable by the plugin itself. For example P003_pulse and P082_cron use this.
This will also be applied in the `formatUserVar` calls, so the value of `[cron#lastexecution]` (or how it is called) will now return the string formatting as shown in the devices tab.
Also the Javascript updating these values will now update the pulse task values and cron task values.  (for pulse there was also some issue about this, have to find and link it)

clumsy-stefan commented:
Great! I will do a new test build later tonight and try it on a few nodes.

Wit "erratic behaviour" you mean the behaviour of the cron plugin and not the (pcf)gpio problem, I assume?

clumsy-stefan commented:
Tried it on 2 nodes:
Cron issue even worse, now it creates an event every second again and never matches... also time calculatiosn seem to be completely wrong:
![image](https://user-images.githubusercontent.com/32835100/107125171-8a5bd280-68a8-11eb-8f82-affbd37a23a2.png)


pcfgpio issues still the same, can't be set anymore..

sorry...

TD-er commented:
I tested with this cron expression: `*/14 * * * * *` (every 14 seconds on :00, :14, :28, :42, :56) and that one seemed fine.

I now have again an internet connection, so I can perform some tests myself.
Not changed anything on the PCF, so if that's still not working, that was to be expected.

clumsy-stefan commented:
probably becuase you use a platform with 32bit time_t and mine is 64bit... can't imagine what else it could be..

TD-er commented:
My changes should take care of just that.

clumsy-stefan commented:
Somehow they didn't ;) could be an alignment problem though... I just it's interesting, that the event is fired every second in my build...

TD-er commented:
Well based on the fact the time is also off by a few months, I guess you can expect just about anything
I will make a different conversion of this and make sure it will not be build platform dependent.
And I will go through all of the code to find all `union` occurrences as this may be a source of "one build works just fine, next failed"

clumsy-stefan commented:
probably time_t changed in later versions from 32 to 64 bit to circumvent certain overflow issues. 32bit Unix-time only works until Jan. 19th 2038 (starting from 1970). Probably that's why it's changed in later frameworks to 64bit (I'm using latest esp8266 core and avr compiler).

TD-er commented:
I know why they may change the `time_t` size, but what you're seeing is something shifted like 24 bits judging by the date shown on your screenshot.
So it looks like an alignment issue.

TD-er committed:
Merge branch 'mega' into bugfix/tm_mon_offset

TD-er commented:
@clumsy-stefan I made a test build.
Can you test if this one does work on your system?
- [ESPEasy_ESP82xx_mega-20210114-55-PR_3487.zip](https://www.dropbox.com/s/m855mi5537ic0yh/ESPEasy_ESP82xx_mega-20210114-55-PR_3487.zip?dl=0)
- [ESPEasy_ESP32_mega-20210114-55-PR_3487.zip](https://www.dropbox.com/s/vu2kuwcybat2b5y/ESPEasy_ESP32_mega-20210114-55-PR_3487.zip?dl=0)

If it does, then we know it isn't a logic bug in the recent changes, but rather some compiling issue as you also have those very strange issues with the cron plugin.

clumsy-stefan commented:
What's in this build? I always use 24tasks build (which are mostly full), ist this also a 24 task build? Also I'm not sure what happens as I have a different set of plugins, so most likely my units will not work with this? 

EDIT: I could use a new blank unit, but there I have no PCF attached... but which .bin in this zip file should I take?

EDIT 2: I'm almost sure the cron stuff is a compiler issue, as I use a much newer version of compiler and esp8266 core than PIO, it already has 64bit time_t...

TD-er commented:
Hmm I can make a 24 task build, if you tell me which one of the 24 task options you use as I remember there were 2 flavours for it.

TD-er commented:
> EDIT 2: I'm almost sure the cron stuff is a compiler issue, as I use a much newer version of compiler and esp8266 core than PIO, it already has 64bit time_t...

I am sure it is a compiler issue, but I want to make sure the PCF stuff you reported is also a compiler issue or a logic error.

clumsy-stefan commented:
> I am sure it is a compiler issue, but I want to make sure the PCF stuff you reported is also a compiler issue or a logic error.

hmm.. but then the issue wouldn't go away after reverting that one commit which brakes it...

I use `#define USE_NON_STANDARD_24_TASKS` in my custom.h and a complete different set of controllers and plugins..

EDIT I have a test-unit with a pcf attached, so if I ccould get a 24taks version I could test without reconfiguring everything...

TD-er commented:
> hmm.. but then the issue wouldn't go away after reverting that one commit which brakes it...

Nope. If it is an alignment issue, you could make it "work" again by adding a longer string to some log item (not reproducible, as you have no guarantee in which order objects are linked during compile)

TD-er committed:
Move to UserVar struct to have universal conversion of task variable types

TD-er commented:
I created a new object for the `UserVar`, which also can perform the conversions when needed.
I tried first with `reinterpret_cast`, but that yielded warnings suggesting there may be an alignment issue (which is what I expect your build may experience)
So now I use a `memcpy` to perform the conversions, which should make sure it is completely build platform independent.

clumsy-stefan commented:
Ok, I can try this tonight. Do I need to make any changes to the playground plugins I use with my builds due to the `UserVar` changes?

TD-er commented:
Nope I used the `operator[]` in the struct to make sure the current code still would work just fine.

clumsy-stefan commented:
Ok, so I did a few tests with the latest code...

The good:
cron seems to work fine now, no strange log entries and the event is called as specified (for the tests I did), especially the "once every minute" seems to work fine!
```
49669: EVENT: Cron#Crontest
49703: ACT : event,date:2021-02-08 21:49:14
49705: EVENT: date:2021-02-08 21:49:14
60806: EVENT: Crontest#LastExecution=2021-02-08 21:49:14
60899: EVENT: Crontest#NextExecution=2021-02-08 21:50:14
```

The bad:
pcfgpio still does not work. it always set's the port to "1" no matter what I give as input.
```
974030: HTTP: pcfgpio,1,0
974033: PCF : port#1: set to 1
````
Shows the same in the return values:
![image](https://user-images.githubusercontent.com/32835100/107278373-ff203f80-6a55-11eb-8d8d-5328d99be196.png)


And another strange thing in the Pulse plugin now, in the log the values it shows are wrong. eg. vor Count it always shows "0" even though in the web-interface it shows the correct value. But only the first one, the other two seem to be fine.
Log:
```
1102045: EVENT: rpm#Count=0
1102148: EVENT: rpm#Total=88087
1102241: EVENT: rpm#Time=27.73
```
Web:
![image](https://user-images.githubusercontent.com/32835100/107278536-37c01900-6a56-11eb-98d7-4dbe6ced9cda.png)
Also the formula used in the device-settings are not executed/calculated anymore now.

TD-er commented:
Good to know the cron is now working fine.
About the PCF state, I just realized it could be a similar issue as you had with cron as the GPIO state struct also does non standard things to keep memory usage low.
When did you upgrade your build env?

I will check the code for generating events to see if I can explain why it may be 0 on the 1st variable.

Formula could somewhat be related to another reported issue where `{D}` is no longer working on plugins like Framed OLED.
No idea yet what to think of that issue.

clumsy-stefan commented:
> Good to know the cron is now working fine.

> About the PCF state, I just realized it could be a similar issue as you had with cron as the GPIO state struct also does non standard things to keep memory usage low.
could be, yes.But as tested, it's that one commit which makes it fail. So I guess that's where to start looking.

> When did you upgrade your build env?
I regularly update the build env from the esp8266 repository (latest commit).
```
System Libraries: ESP82xx Core 27da3476, NONOS SDK 2.2.2-dev(38a443e), LWIP: 2.1.2
```

> Formula could somewhat be related to another reported issue where `{D}` is no longer working on plugins like Framed OLED.
> No idea yet what to think of that issue.
the forumlas I use are:
```
%value% * 6
60000 / %value%
```
to calculate RPM values from Count and from Time and compare them. Task run's every 10 sec.
The latest commit from here (but with this PR) it still works. latest commit seems to break it.

PS: I'll finish now and go to sleep, I get confused with all the different test-builds I've done, need to sort that out tomorrow when my head's clear again ;)

But thanks a lot for your efforts, I'm happy to continue test, analyse and report!

n8

TD-er commented:
I found why you're seeing the event with the pulse counter value set to 0 in the event.
It may also explain why your formula is no longer working.

What I changed...
There is the UserVar array which contains the float values of task values.
For some plugins, there is a special call from the web interface to allow for plugin specific formatting of a task value.
For example the cron plugin does show a formatted time/date string instead of a float.
What I changed was the `formatUserVar` function is now also using this same plugin specific formatting call so we have the best formatting matching any plugin, but more importantly it is showing the same formatting everywhere.

The Pulse plugin also used this specific call to format the values specific for this plugin.
However it did not just use a specific format, but rather it was not showing values from the UserVar array, but rather the internal counters like `Plugin_003_pulseCounter`
This `Plugin_003_pulseCounter` is cleared each time a `PLUGIN_READ` is called and right after this call an event is generated... with the cleared value.

So this is clearly a bug in the Pulse plugin which does also make clear why we're seeing so many very strange and hard to reproduce issues for this plugin.

I think I will for now revert this change for the Pulse plugin and make a new issue for it as it may involve quite a big change. (which I will explain in that issue)

TD-er committed:
[P003 Pulse] Revert changes (See #3500)

Will be fixed in #3500

clumsy-stefan commented:
Ok, so I made a new version now with this changes reverted. I guess the only issue left is the gpio setting which does not work anymore with commit #8a6dc87b4bda4fe32fec6dedf8911aa063be8f66.

I'll report back if I find any other issues with this PR..

TD-er commented:
Thanks for testing.
And I am glad you found those issues as it helped me to finally understand what was wrong with P003_Pulse.

I will later this evening try to find out what is causing your issues with the PCF addressing. I am fairly confident it is a build issue on your side, but still it needs to be fixed as it is something that we may also encounter later when the toolchain gets updated.

clumsy-stefan commented:
Agree, and thanks for looking into it.

BTW: there is another issue coming up in the new toolchain, waveform handling is beeing changed twice, first via static `#define` value later via system call to define if it's phase locked or not.

TD-er commented:
OK, so I have to define it myself I guess in the PIO config.
For Arduino IDE it has to be defined elsewhere, so I will also need to update that in the docs.

clumsy-stefan commented:
As in the latest version the `#define` is not needed anymore, it's done via system call at runtime, so it has to be specified in the code (I added it in the main .ino file) which variant you want to use.

TD-er commented:
Can you give an example of how to set it?

clumsy-stefan commented:
Sure, in `setup()` you need to call `enablePhaseLockedWaveform()` to use phase lock waveform ( see esp8266 commit #f5fd5912fe09d7eec8a909326287379570579ea6) which is defined in `#include <core_esp8266_waveform.h>`.

TD-er commented:
I will merge this for now and then start to look into your reported issue for the PCF.
I will also add the suggested code for PWM on the latest esp8266/Arduino code base.

clumsy-stefan commented:
> I will merge this for now and then start to look into your reported issue for the PCF.

So I hope I'm the only one experiencing the gpio issues, otherwise a lot of people will be unhappy after the merge ;) I can always revert the relevant commit, as I have my local build..

Other than this I'm running this PR with reverted commit together with PR #3499 for about a day now on all my nodes without any problems!

> I will also add the suggested code for PWM on the latest esp8266/Arduino code base.
Ok, not sure itf it's the most "elegant" way to do it, but it works ;)

TD-er commented:
> So I hope I'm the only one experiencing the gpio issues, otherwise a lot of people will be unhappy after the merge ;) I can always revert the relevant commit, as I have my local build..

"A lot of people" are only running the pre-built images.
Only a handful use the 'mega' branch and this situation will not last for long.

TD-er commented:
As you may have noticed, I merged a few commits, most of them based on what you mentioned.
Can you test the current 'mega' branch?

clumsy-stefan commented:
Sure, I can do a test build later today. But I wouldn't expect any issues, as this seems to be more or less what I've running for 2 days now without issues.

I'll report back probably tonight, latest tomorrow!

Why did you revert the PCF commit now? Any reports from others that it doesn't work?

TD-er commented:
> Why did you revert the PCF commit now? Any reports from others that it doesn't work?

There were some reports of issues which I could not explain, but also your reported issue I cannot explain based on what has been changed in that commit.
Since the new functionality added in that commit was not yet used and it is unclear whether it is buggy or not, I better go for the route we know is working.
I will later have another look at that code to see what may be wrong here.
But for now I really would like to move on to a new nightly build so we can finally work towards a release to make sure people don't start with build R120 from 4 years ago as this mega branch doesn't have a "release" label on it.

clumsy-stefan commented:
Tested it on a handful of nodes and the main issues (PCF and Cron) seem to be solved, very nice, thanks. I think this could go into the standard build.

But due to the a bit more intensive testing, I encountered different things, especially with the Pulse Plugin. I'll comment that on #3499, as I guess they are independent from this change.

One thing though which I see since a few weeks, are random WD reboots with following infos after reboot:
```
Boot: Manual reboot (2)
Reset Reason: Software/System restart
Last Action before Reboot: Const Interval: TIMER_C006_DELAY_QUEUE
```
What looks especially strange to me, is that I don't even include C006 in my build, so why does it use it?

Sometimes also an exception in PLUGIN_READ....

but can't really comment on this further, need to do more tests...

TD-er commented:
These units that crash have all the pulse plugin active?

clumsy-stefan commented:
All of the units have the plugin compiled in. only 2 of them uses it actually. But the issues seems to occur randomly... Sometimes also the unit with the pulsecounter runs vor 2 days without issues and then reboots... so I guess this is a complete different issue...
